crates/admin/src/cluster_controller/logs_controller/nodeset_selection.rs

```diff
@@ -127,12 +164,16 @@ impl<'a> NodeSetSelector<'a> {
             }
         };
 
+        // even with all possible dead node we still can't reach the optimal
+        // nodeset size. This means there are not enough nodes in the cluster
+        // we still return an error.
+
         // todo: implement location scope-aware selection
-        if nodeset.len() < nodeset_min_size {
+        if nodeset.len() < nodeset_size.optimal_size {
             trace!(
-                "Failed to place replicated loglet: insufficient writeable nodes to meet minimum size requirement {} < {}",
+                "Failed to place replicated loglet: insufficient writeable nodes to meet optimal size requirement {} < {}",
```

Comment:
If we can't have optimal size it means there are not enough nodes in the cluster to begin with. It's allowed to have `minimum` alive node, as long it has enough nodes to satisfy the `optimal` size (even if those nodes are dead). This allows us to wait until the cluster has enough nodes before it can provision a replicated loglet but does not stop it from reconfiguration if one node is dead (in a 3 node cluster)

So what happens is the follows:
- try to find "target size" of the nodes from the alive AND preferred nodeset
- If target size was not satisfied, we try to fill it up from any other alive node
- if the selected nodeset can not satisfy the minimum size that's an error since we can't have a write quorum and the Insufficient Nodes error is returned.
- Otherwise we try to fill up the `optimal` size from the remaining writable (but dead) nodes, again we prioritise preferred nodes. This MUST satisfy the optimal size but if this doesn't happen it means there are no enough nodes in the cluster and even with enough nodes to satisfy the minumum requirements we should not proceed. This allows the cluster to wait until enough nodes are up.


I hope this makes it clear. If you think there is a better way to rephrase the log line to make it clearer please let me know.